lib/test.js

```diff
@@ -550,14 +551,28 @@ Test.prototype._runChild = function runChild (child, name, extra, cb) {
 
   // still need try/catch for synchronous errors
   try {
+    child._mustDeferEnd = true;
     var cbRet = cb(child)
     if (cbRet && typeof cbRet.then === 'function') {
       // promise
       cbRet.then(function () {
-        child.end(IMPLICIT)
-      }, child.threw)
+        child._mustDeferEnd = false;
+        if (!child._tryResumeEnd()) {
+          child.end(IMPLICIT)
+        }
+      }, function (err) {
+        child._mustDeferEnd = false;
+        if (!child._tryResumeEnd()) {
```

Comment:
@rmg I'm not quite sure, it's been a while since I wrote the code here 😄 

If `cb(child)` throws then this code is never executed. What could happen is that the test is failed and then the promise is rejected. I suppose this would then prevent the promise rejection from being logged.

Possibly we should `_tryResumeEnd()` and regardless of its return value still run `child.threw(err)`. But it sounds like we'd need a test case for this behavior first.